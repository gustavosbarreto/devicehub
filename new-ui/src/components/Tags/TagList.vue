<template>
  <v-table v-bind="$attrs" data-test="tagListList-dataTable" class="bg-v-theme-surface">
    <thead>
      <tr>
        <th
          v-for="(head, i) in headers"
          :key="i"
          :class="head.align ? `text-${head.align}` : 'text-center'"
        >
          <span> {{ head.text }}</span>
        </th>
      </tr>
    </thead>
    <tbody v-if="tags.length">
      <tr v-for="(tag, i) in tags" :key="i">
        <td class="text-center">{{ tag }}</td>
        <td class="text-center">
          <v-menu location="bottom" scrim eager>
            <template v-slot:activator="{ props }">
              <v-chip v-bind="props" density="comfortable" size="small">
                <v-icon>mdi-dots-horizontal</v-icon>
              </v-chip>
            </template>
            <v-list class="bg-v-theme-surface" lines="two" density="compact">
              <v-tooltip location="bottom" class="text-center" :disabled="hasAuthorizationEdit()">
                <template v-slot:activator="{ props }">
                  <div v-bind="props">
                    <TagEdit
                      :tag="tag"
                      :not-has-authorization="!hasAuthorizationEdit()"
                      @update="getTags()"
                    />
                  </div>
                </template>
                <span> You don't have this kind of authorization. </span>
              </v-tooltip>

              <v-tooltip location="bottom" class="text-center" :disabled="hasAuthorizationRemove()">
                <template v-slot:activator="{ props }">
                  <div v-bind="props">
                    <TagRemove
                      :tag="tag"
                      :not-has-authorization="!hasAuthorizationRemove()"
                      @update="getTags()"
                    />
                  </div>
                </template>
                <span> You don't have this kind of authorization. </span>
              </v-tooltip>
            </v-list>
          </v-menu>
        </td>
      </tr>
    </tbody>
    <div v-else class="text-start mt-2 text-medium-emphasis">
      <p>No data avaliabe</p>
    </div>
  </v-table>
</template>

<script lang="ts">
import { defineComponent, computed, onMounted } from "vue";
import { useStore } from "../../store";
import { actions, authorizer } from "../../authorizer";
import hasPermission from "../../utils/permission";
import TagRemove from "./TagRemove.vue";
import TagEdit from "./TagEdit.vue";
import { INotificationsError } from "../../interfaces/INotifications";

export default defineComponent({
  inheritAttrs: true,
  setup() {
    const store = useStore();
    const tags = computed(() => store.getters["tags/list"]);
    const hasAuthorizationEdit = () => {
      const role = store.getters["auth/role"];
      if (role !== "") {
        return hasPermission(authorizer.role[role], actions.tag.edit);
      }
      return false;
    };
    const hasAuthorizationRemove = () => {
      const role = store.getters["auth/role"];
      if (role !== "") {
        return hasPermission(authorizer.role[role], actions.tag.remove);
      }
      return false;
    };
    const getTags = async () => {
      try {
        await store.dispatch("tags/fetch");
      } catch (error: any) {
        store.dispatch(
          "snackbar/showSnackbarErrorLoading",
          INotificationsError.deviceTagList,
        );
        throw new Error(error);
      }
    };
    onMounted(() => {
      getTags();
    });
    const updateTags = () => {
      getTags();
    };
    return {
      headers: [
        {
          text: "Name",
          value: "name",
          align: "center",
          sortable: false,
        },
        {
          text: "Actions",
          value: "actions",
          align: "center",
          sortable: false,
        },
      ],
      tags,
      hasAuthorizationEdit,
      hasAuthorizationRemove,
      updateTags,
      getTags,
    };
  },
  components: { TagRemove, TagEdit },
});
</script>
