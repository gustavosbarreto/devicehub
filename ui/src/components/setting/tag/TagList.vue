<template>
  <fragment>
    <v-data-table
      :headers="headers"
      :items="getListTags"
      :server-items-length="getNumberTags"
      hide-default-footer
      data-test="tagListList-dataTable"
    >
      <template #[`item.name`]="{ item }">
        {{ item.name }}
      </template>

      <template #[`item.actions`]="{ item }">
        <TagFormDialog
          action="edit"
          :tag-name="item.name"
          data-test="tagFormDialog-component"
          @update="getTags()"
        />

        <TagDelete
          :tag-name="item.name"
          data-test="tagDelete-component"
          @update="getTags()"
        />
      </template>
    </v-data-table>
  </fragment>
</template>

<script>

import TagFormDialog from '@/components/setting/tag/TagFormDialog';
import TagDelete from '@/components/setting/tag/TagDelete';

export default {
  name: 'TagList',

  components: {
    TagFormDialog,
    TagDelete,
  },

  data() {
    return {
      headers: [
        {
          text: 'Name',
          value: 'name',
          align: 'center',
        },
        {
          text: 'Actions',
          value: 'actions',
          align: 'center',
          sortable: false,
        },
      ],
    };
  },

  computed: {
    getListTags() {
      return this.$store.getters['tags/list'].map((str) => ({ name: str }));
    },

    getNumberTags() {
      return this.$store.getters['tags/getNumberTags'];
    },
  },

  created() {
    this.getTags();
  },

  methods: {
    async getTags() {
      try {
        await this.$store.dispatch('tags/fetch');
      } catch (error) {
        this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.snackbar.deviceTagList);
      }
    },
  },
};
</script>
