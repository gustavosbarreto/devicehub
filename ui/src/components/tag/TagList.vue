<template>
  <fragment>
    <v-data-table
      :headers="headers"
      :items="getListTags"
      :server-items-length="getNumberTags"
      hide-default-footer
      data-test="tagListList-dataTable"
    >
      <template #[`item.name`]="{ item }">
        {{ item.name }}
      </template>

      <template #[`item.actions`]="{ item }">
        <v-menu
          :ref="'menu'+getListTags.indexOf(item)"
          offset-y
        >
          <template #activator="{ on, attrs }">
            <v-chip
              color="transparent"
              v-on="on"
            >
              <v-icon
                small
                class="icons"
                v-bind="attrs"
                v-on="on"
              >
                mdi-dots-horizontal
              </v-icon>
            </v-chip>
          </template>

          <v-card>
            <v-tooltip
              bottom
              :disabled="hasAuthorizationEdit"
            >
              <template #activator="{ on, attrs }">
                <div
                  v-bind="attrs"
                  v-on="on"
                >
                  <v-list-item
                    :disabled="!hasAuthorizationEdit"
                    @click="showTagDialog(getListTags.indexOf(item))"
                  >
                    <TagFormDialogEdit
                      :tag-name="item.name"
                      :show.sync="tagDialogShow[getListTags.indexOf(item)]"
                      data-test="tagFormDialogEdit-component"
                      @update="getTags()"
                    />
                  </v-list-item>
                </div>
              </template>

              <span>
                You don't have this kind of authorization.
              </span>
            </v-tooltip>

            <v-tooltip
              bottom
              :disabled="hasAuthorizationRemove"
            >
              <template #activator="{ on, attrs }">
                <div
                  v-bind="attrs"
                  v-on="on"
                >
                  <v-list-item
                    :disabled="!hasAuthorizationRemove"
                    @click="showTagDelete(getListTags.indexOf(item))"
                  >
                    <TagDelete
                      :tag-name="item.name"
                      :show.sync="tagDeleteShow[getListTags.indexOf(item)]"
                      data-test="tagDelete-component"
                      @update="getTags()"
                    />
                  </v-list-item>
                </div>
              </template>

              <span>
                You don't have this kind of authorization.
              </span>
            </v-tooltip>
          </v-card>
        </v-menu>
      </template>
    </v-data-table>
  </fragment>
</template>

<script>

import TagFormDialogEdit from '@/components/tag/TagFormDialogEdit';
import TagDelete from '@/components/tag/TagDelete';

import hasPermission from '@/components/filter/permission';

export default {
  name: 'TagListComponent',

  filters: { hasPermission },

  components: {
    TagFormDialogEdit,
    TagDelete,
  },

  data() {
    return {
      tagDialogShow: [],
      tagDeleteShow: [],
      editAction: 'edit',
      removeAction: 'remove',
      headers: [
        {
          text: 'Name',
          value: 'name',
          align: 'center',
        },
        {
          text: 'Actions',
          value: 'actions',
          align: 'center',
          sortable: false,
        },
      ],
    };
  },

  computed: {
    getListTags() {
      return this.$store.getters['tags/list'].map((str) => ({ name: str }));
    },

    getNumberTags() {
      return this.$store.getters['tags/getNumberTags'];
    },

    hasAuthorizationEdit() {
      const role = this.$store.getters['auth/role'];
      if (role !== '') {
        return hasPermission(
          this.$authorizer.role[role],
          this.$actions.tag[this.editAction],
        );
      }

      return false;
    },

    hasAuthorizationRemove() {
      const role = this.$store.getters['auth/role'];
      if (role !== '') {
        return hasPermission(
          this.$authorizer.role[role],
          this.$actions.tag[this.removeAction],
        );
      }

      return false;
    },
  },

  created() {
    this.getTags();
  },

  methods: {
    async getTags() {
      try {
        await this.$store.dispatch('tags/fetch');

        this.setArrays();
      } catch (error) {
        this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.snackbar.deviceTagList);
      }
    },

    showTagDialog(index) {
      this.tagDialogShow[index] = this.tagDialogShow[index] === undefined
        ? true : !this.tagDialogShow[index];
      this.$set(this.tagDialogShow, index, this.tagDialogShow[index]);

      this.closeMenu(index);
    },

    showTagDelete(index) {
      this.tagDeleteShow[index] = this.tagDeleteShow[index] === undefined
        ? true : !this.tagDeleteShow[index];
      this.$set(this.tagDeleteShow, index, this.tagDeleteShow[index]);

      this.closeMenu(index);
    },

    setArrays() {
      const numberTags = this.getListTags.length;

      if (numberTags > 0) {
        this.tagDialogShow = new Array(numberTags).fill(false);
        this.tagDeleteShow = new Array(numberTags).fill(false);
      }
    },

    closeMenu(index) {
      this.$refs[`menu${index}`].isActive = false;
    },
  },
};
</script>
