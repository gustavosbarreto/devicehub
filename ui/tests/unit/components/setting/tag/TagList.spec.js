import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import TagList from '@/components/setting/tag/TagList';

describe('TagList', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();

  localVue.use(Vuex);

  let wrapper;

  const status = true;
  const numberTags = 3;
  const tags = ['tag1', 'tag2', 'tag3'];
  const tagsObject = tags.map((str) => ({ name: str }));

  const headers = [
    {
      text: 'Name',
      value: 'name',
      align: 'center',
    },
    {
      text: 'Actions',
      value: 'actions',
      align: 'center',
      sortable: false,
    },
  ];

  const storeWithoutData = new Vuex.Store({
    namespaced: true,
    state: {
      tags: [],
      numberTags: 0,
      status,
    },
    getters: {
      'tags/list': (state) => state.tags,
      'tags/getNumberTags': (state) => state.numberTags,
    },
    actions: {
      'tags/fetch': () => {},
      'snackbar/showSnackbarErrorLoading': () => {},
    },
  });

  const storeWithData = new Vuex.Store({
    namespaced: true,
    state: {
      tags,
      numberTags,
      status,
    },
    getters: {
      'tags/list': (state) => state.tags,
      'tags/getNumberTags': (state) => state.numberTags,
    },
    actions: {
      'tags/fetch': () => {},
      'snackbar/showSnackbarErrorLoading': () => {},
    },
  });

  describe('Without Data', () => {
    beforeEach(() => {
      wrapper = mount(TagList, {
        store: storeWithoutData,
        localVue,
        stubs: ['fragment', 'router-link'],
        mocks: {
          $env: (isEnterprise) => isEnterprise,
        },
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.headers).toEqual(headers);
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.getListTags).toEqual([]);
      expect(wrapper.vm.getNumberTags).toEqual(0);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with components', () => {
      expect(wrapper.find('[data-test="tagFormDialog-component"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="tagDelete-component"]').exists()).toBe(false);
    });
    it('Renders the template with data', () => {
      const dt = wrapper.find('[data-test="tagListList-dataTable"]');
      const dataTableProps = dt.vm.$options.propsData;

      expect(dataTableProps.items).toHaveLength(0);
    });
  });

  describe('With Data', () => {
    beforeEach(() => {
      wrapper = mount(TagList, {
        store: storeWithData,
        localVue,
        stubs: ['fragment', 'router-link'],
        mocks: {
          $env: (isEnterprise) => isEnterprise,
        },
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.headers).toEqual(headers);
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.getListTags).toEqual(tagsObject);
      expect(wrapper.vm.getNumberTags).toEqual(numberTags);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with components', () => {
      expect(wrapper.find('[data-test="tagFormDialog-component"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="tagDelete-component"]').exists()).toBe(true);
    });
    it('Renders the template with data', () => {
      const dt = wrapper.find('[data-test="tagListList-dataTable"]');
      const dataTableProps = dt.vm.$options.propsData;

      expect(dataTableProps.items).toHaveLength(numberTags);
    });
  });
});
