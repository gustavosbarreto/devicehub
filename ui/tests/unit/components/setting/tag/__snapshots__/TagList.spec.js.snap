// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagList With Data Renders the component 1`] = `
<fragment-stub>
  <div class="v-data-table theme--light" data-test="tagListList-dataTable">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Name: Not sorted. Activate to sort ascending." aria-sort="none" class="text-center sortable"><span>Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Actions" class="text-center"><span>Actions</span></th>
          </tr>
        </thead>
        <tbody>
          <tr class="">
            <td class="text-center">
              tag1
            </td>
            <td class="text-center">
              <fragment-stub data-test="tagFormDialog-component"><span><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">edit</button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
                <div role="dialog" class="v-dialog__container">
                  <!---->
                </div>
              </fragment-stub>
              <fragment-stub data-test="tagDelete-component"><span><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">delete</button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
                <div role="dialog" class="v-dialog__container">
                  <!---->
                </div>
              </fragment-stub>
            </td>
          </tr>
          <tr class="">
            <td class="text-center">
              tag2
            </td>
            <td class="text-center">
              <fragment-stub data-test="tagFormDialog-component"><span><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">edit</button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
                <div role="dialog" class="v-dialog__container">
                  <!---->
                </div>
              </fragment-stub>
              <fragment-stub data-test="tagDelete-component"><span><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">delete</button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
                <div role="dialog" class="v-dialog__container">
                  <!---->
                </div>
              </fragment-stub>
            </td>
          </tr>
          <tr class="">
            <td class="text-center">
              tag3
            </td>
            <td class="text-center">
              <fragment-stub data-test="tagFormDialog-component"><span><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">edit</button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
                <div role="dialog" class="v-dialog__container">
                  <!---->
                </div>
              </fragment-stub>
              <fragment-stub data-test="tagDelete-component"><span><button disabled="disabled" type="button" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">delete</button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
                <div role="dialog" class="v-dialog__container">
                  <!---->
                </div>
              </fragment-stub>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</fragment-stub>
`;

exports[`TagList Without Data Renders the component 1`] = `
<fragment-stub>
  <div class="v-data-table theme--light" data-test="tagListList-dataTable">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Name: Not sorted. Activate to sort ascending." aria-sort="none" class="text-center sortable"><span>Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Actions" class="text-center"><span>Actions</span></th>
          </tr>
        </thead>
        <tbody>
          <tr class="v-data-table__empty-wrapper">
            <td colspan="2">No data available</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</fragment-stub>
`;
